rust/agama-lib/src/utils/transfer/file_systems.rs

```diff
@@ -0,0 +1,284 @@
+}
+
+/// Holds a list of file systems.
+///
+/// It offers a set of convenience method to search within the list.
+#[derive(Debug, Default)]
+pub struct FileSystemsList {
+    file_systems: Vec<FileSystem>,
+}
+
+impl FileSystemsList {
+    /// Creates a list of file systems.
+    pub fn new(file_systems: Vec<FileSystem>) -> Self {
+        Self { file_systems }
+    }
+
+    /// Creates a list for the file systems in the underlying system.
+    pub fn from_system() -> Self {
+        let file_systems = FileSystemsReader::read_from_system();
+        Self::new(file_systems)
+    }
+
+    pub fn to_vec(&self) -> Vec<FileSystem> {
+        self.file_systems.clone()
+    }
+
+    /// Returns the file system with the given block device name.
+    ///
+    /// * `name`: block device name.
+    pub fn by_name(&self, name: &str) -> Option<&FileSystem> {
+        self.file_systems.iter().find(|fs| name == &fs.block_device)
+    }
+
+    /// Returns the file systems with the given label name.
+    ///
+    /// * `label`: device label.
+    pub fn by_label(&mut self, label: &str) -> Self {
+        let label = Some(label.to_string());
+        let file_systems = self
+            .file_systems
+            .iter()
+            .filter(|fs| fs.label == label)
+            .cloned()
+            .collect();
+
+        FileSystemsList { file_systems }
+    }
+
+    /// Returns the file systems using the given transport.
+    ///
+    /// * `transport`: transport of the device (e.g., "usb").
+    pub fn by_transport(&mut self, transport: &str) -> Self {
+        let transport = Some(transport.to_string());
+        let file_systems = self
+            .file_systems
+            .iter()
+            .filter(|fs| fs.transport == transport)
+            .cloned()
+            .collect();
+
+        FileSystemsList { file_systems }
+    }
+}
+
+/// Implements the logic to read the file systems from the underlying system.
+///
+/// This struct relies on lsblk to find the file systems. It is extracted to
+/// a separate struct to make testing easier.
+struct FileSystemsReader {}
+
+impl FileSystemsReader {
+    /// Returns the file systems from the underlying system.
+    pub fn read_from_system() -> Vec<FileSystem> {
+        let lsblk = Command::new("lsblk")
+            .args([
+                "--output",
+                "KNAME,FSTYPE,MOUNTPOINTS,TRAN,LABEL",
+                "--pairs",
+                "--path",
+            ])
+            .output()
+            .unwrap();
+        let output = String::from_utf8_lossy(&lsblk.stdout);
+        Self::read_from_string(&output)
+    }
+
+    /// Turns the output of lsblk into a list of file systems.
+    pub fn read_from_string(lsblk_string: &str) -> Vec<FileSystem> {
+        let mut file_systems = vec![];
+        let mut parent_transport: Option<String> = None;
+        let re =
+            Regex::new(r#"KNAME="(.+)" FSTYPE="(.*)" MOUNTPOINTS="(.*)" TRAN="(.*)" LABEL="(.*)""#)
+                .unwrap();
```

Comment:
yeah, that unwrap for regexp is what I hate on regexp API. Ideally it should be checked during compilation for static strings and avoid those runtime unwraps.